File: src/features/typing/data/ts/implement_trie_prefix_tree.ts
class Trie {
    private children: (Trie | null)[];
    private isEnd: boolean;

    constructor() {
        this.children = new Array(26).fill(null);
        this.isEnd = false;
    }

    insert(word: string): void {
        let node: Trie = this;
        for (const ch of word) {
            const idx = ch.charCodeAt(0) - 97;
            if (node.children[idx] === null) {
                node.children[idx] = new Trie();
            }
            node = node.children[idx]!;
        }
        node.isEnd = true;
    }

    search(word: string): boolean {
        const node = this.searchPrefix(word);
        return node !== null && node.isEnd;
    }

    startsWith(prefix: string): boolean {
        return this.searchPrefix(prefix) !== null;
    }

    private searchPrefix(prefix: string): Trie | null {
        let node: Trie | null = this;
        for (const ch of prefix) {
            node = node.children[ch.charCodeAt(0) - 97];
            if (node === null) {
                return null;
            }
        }
        return node;
    }
}

File: src/features/typing/data/go/implement_trie_prefix_tree.go
package main

type Trie struct {
	children [26]*Trie
	isEnd    bool
}

func NewTrie() Trie {
	return Trie{}
}

func (t *Trie) Insert(word string) {
	node := t
	for _, ch := range word {
		idx := ch - 'a'
		if node.children[idx] == nil {
			node.children[idx] = &Trie{}
		}
		node = node.children[idx]
	}
	node.isEnd = true
}

func (t *Trie) Search(word string) bool {
	node := t.searchPrefix(word)
	return node != nil && node.isEnd
}

func (t *Trie) StartsWith(prefix string) bool {
	return t.searchPrefix(prefix) != nil
}

func (t *Trie) searchPrefix(prefix string) *Trie {
	node := t
	for _, ch := range prefix {
		node = node.children[ch-'a']
		if node == nil {
			return nil
		}
	}
	return node
}
